### ISSUE TITLE
DNS lookup fails on statefulset pod when spawned after init-container

### ISSUE THREAD

**Issue by prat0318** [kubernetes/kubernetes#60714]
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**: When kafka statefulset cluster (in conjunction with initContainers) is rolled, it is not able to do DNS lookups. Repeated restarts on the same node fail repeatedly. Only when I `cordon` that node, and delete the kafka pod, and kafka gets reassigned to another node, it is able to make DNS lookup successfully (even when initContainer is running along with it).

**What you expected to happen**: I expected the DNS lookups to work just fine on the statefulset pods after rolling restart.

**How to reproduce it (as minimally and precisely as possible)**:
* Add a initContainer [like so](https://github.com/kubernetes/kubernetes/issues/44103#issuecomment-326851272)
```
      initContainers:
      - name: label-set
        command: ["sh","-c","/usr/local/bin/kubectl label pod ${HOSTNAME} podname=${HOSTNAME} -n{{ .Values.name }}"]
        image: my.private.registry/k8s/kubectl
```
* Roll the statefulset by changing an annotation
* Ensure that the pod again comes up on the last node it was running on.
* Verify that the `kafka` pod fails to start up:
```
Caused by: java.net.UnknownHostException: zookeeper-service.zookeeper.svc.cluster.local: Temporary failure in name resolution\n","stream":"stdout","time":"2018-03-02T12:36:58.167197901Z
```
* Confirm zookeeper-service is running:
```
$ kubectl -nzookeeper get svc
NAME                CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
zookeeper           None            <none>        2888/TCP,3888/TCP   72d
zookeeper-service   172.20.224.45   <none>        2181/TCP            72d
```
----
Normal Scenario
```
$ kubectl -nkafka exec -it kafka-0 bash
# telnet zookeeper-service.zookeeper.svc.cluster.local 2181
Trying 172.20.224.45...
Connected to zookeeper-service.zookeeper.svc.cluster.local.
Escape character is '^]'.
```
When deleted the pod, and it again lands up on the same node:
```
$ kubectl -nkafka delete po kafka-0
pod "kafka-0" deleted
$kubectl -nkafka get po -o wide -w
NAME      READY     STATUS        RESTARTS   AGE       IP               NODE
kafka-0   1/1       Terminating   0          1m        172.16.234.231   kubelarge06
kafka-1   1/1       Running       0          2h        172.16.184.167   kubelarge07
kafka-2   1/1       Running       0          2h        172.16.77.117    kubelarge08
kafka-0   0/1       Terminating   0         1m        <none>    kubelarge06
kafka-0   0/1       Terminating   0         1m        <none>    kubelarge06
kafka-0   0/1       Terminating   0         1m        <none>    kubelarge06
kafka-0   0/1       Pending   0         36s       <none>    <none>
kafka-0   0/1       Pending   0         36s       <none>    kubelarge06
kafka-0   0/1       Init:0/1   0         36s       <none>    kubelarge06
kafka-0   0/1       Init:0/1   0         39s       172.16.234.250   kubelarge06
kafka-0   0/1       Init:0/1   0         40s       172.16.234.250   kubelarge06
kafka-0   0/1       PodInitializing   0         40s       172.16.234.250   kubelarge06
kafka-0   1/1       Running   0         41s       172.16.234.250   kubelarge06
$ kubectl -nkafka exec -it kafka-0 bash
# telnet zookeeper-service.zookeeper.svc.cluster.local 2181
telnet: could not resolve zookeeper-service.zookeeper.svc.cluster.local/2181: Temporary failure in name resolution
```
Lets cordon `kubelarge06`:
```
$ kubectl cordon kubelarge06
node "kubelarge06" cordoned
$ kubectl -nkafka delete po kafka-0
pod "kafka-0" deleted
$ kubectl -nkafka get po -o wide -w
NAME      READY     STATUS        RESTARTS   AGE       IP               NODE
kafka-0   0/1       Terminating   4          6m        <none>           kubelarge06
kafka-1   1/1       Running       0          3h        172.16.184.167   kubelarge07
kafka-2   1/1       Running       0          3h        172.16.77.117    kubelarge08
kafka-0   0/1       Terminating   4         6m        <none>    kubelarge06
kafka-0   0/1       Terminating   4         6m        <none>    kubelarge06
kafka-0   0/1       Pending   0         37s       <none>    <none>
kafka-0   0/1       Pending   0         37s       <none>    kubemedium09
kafka-0   0/1       Init:0/1   0         37s       <none>    kubemedium09
kafka-0   0/1       Init:0/1   0         40s       172.16.6.90   kubemedium09
kafka-0   0/1       Init:0/1   0         40s       172.16.6.90   kubemedium09
kafka-0   0/1       PodInitializing   0         41s       172.16.6.90   kubemedium09
kafka-0   1/1       Running   0         42s       172.16.6.90   kubemedium09

$ kubectl -nkafka exec -it kafka-0 bash
root@kafka-0:/usr/java# telnet zookeeper-service.zookeeper.svc.cluster.local 2181
Trying 172.20.224.45...
Connected to zookeeper-service.zookeeper.svc.cluster.local.
Escape character is '^]'.
^]
telnet>
```

**Anything else we need to know?**: It works fine when scheduled again on another node than the one previously running.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T07:00:21Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T06:43:48Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Onpremise VMs
- OS (e.g. from /etc/os-release):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`): Linux bla 3.10.0-693.11.6.el7.x86_64 #1 SMP Thu Jan 4 01:06:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kismatic
- Others:
  CNI: calico

**DNS Configuration:**
```
$ kubectl -nkafka exec -it kafka-0 bash
root@kafka-0:/usr/java# cat /etc/resolv.conf
nameserver 172.20.0.2
search kafka.svc.cluster.local svc.cluster.local cluster.local
options ndots:5

$ kubectl -nkube-system get svc
NAME                   CLUSTER-IP       EXTERNAL-IP   PORT(S)         AGE
heapster               172.20.63.41     <none>        80/TCP          154d
heapster-influxdb      172.20.167.202   <none>        8086/TCP        154d
kube-dns               172.20.0.2       <none>        53/UDP,53/TCP   154d
kubernetes-dashboard   172.20.113.235   <none>        80/TCP          154d
tiller-deploy          172.20.80.89     <none>        44134/TCP       154d
```

**Comment by prat0318** [comment]
@kubernetes/sig-network-bug

**Comment by prat0318** [comment]
@kubernetes/sig-network-bugs

**Comment by rocketraman** [comment]
I believe I just ran into this same issue. I don't have an init container, but upon startup when Kafka attempted to reconnect to zookeeper, the restarted pod in the stateful set was unable to connect. Resolution to zookeeper was checked from the other pods still running, via `exec` + `nslookup` and they had no issues resolving the name.

The OPs workaround of cordoning the host leads me to believe that something was wrong with the DNS resolution on the host itself. So I flushed the arp cache and the local DNS resolver cache via:

```
ip -s -s neigh flush all
kill -HUP $(pidof /usr/sbin/nscd)
```

and then restarted the container again. This seems to have solved the problem. I don't know why this happens, but it seems like a pretty serious problem -- not sure why this hasn't gotten more attention.
